### ISSUE TITLE
[Feature Request]: tf.data.Dataset.map parallelism autotune enhancement

### ISSUE THREAD

**Issue by fanshiqing** [tensorflow/tensorflow#22465]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Cent OS
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.4
- **Python version**: 2.7.5
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**: NA

### Describe the problem
Recently this [PR](https://github.com/tensorflow/tensorflow/commit/c8a0dfc741736a59f8fd1776b71f38619d66da56#diff-df634c8243713c0afd2e05c1689412e2) which leverages the underlying  performance model to find the optimal values for the parallelism knobs, this seems intuitive and really useful for me. While after carefully going through the relevant codebase, there are two problems that have been bothering me.

(1) Currently dataset.map's num_parallel_calls autotune (as well as [dataset.prefetch buffer size autotune](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/prefetch_autotuner.cc#L36-L55)) only works in the scenario when `kAutotune` guard is set and then start a one-way increment. On the one hand, in actual production environment, careless setting of parameters like `map.num_parllel_calls` or `prefetch.buffer_size` is somewhat common, which may result in inefficient  or even wore resource utilization.  On the other hand, even a carefully-tuned parameter which leverages underlying system info and works good in environment may turn out to be incompatible when we transfer into another environment. Therefore it's nature that our AUTOTUNE logic can support our auto_tune params's autotune based on runtime system info **as well as user's initial params passed in**, in which case the autotune_params' adjustment should be two-way.

(2) At the specific code implementation level, this PR leverages `port::NumSchedulableCPUs()` to get the number of cores available in the process, and calculate the idealized acceleration brought by parallelism's increment(e.g. [PARALLEL_MAP](https://github.com/tensorflow/tensorflow/blob/c8a0dfc741736a59f8fd1776b71f38619d66da56/tensorflow/core/framework/model.cc#L214-L221)). This should be tuned carefully as in complicated distributed environment the increase of map parallelism perhaps results in invalid scheduling of resources, and eventually **brings in negative optimization**.

For the above two points I have figured out some workaround, and the first draft of the code is also being developed. Any comments from you side is hight welcome and much appreciated. @jsimsa 

Thanks

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Mobile device

**Comment by wangsiyu** [comment]
Agree. The auto tuning of `map.num_parallel_calls` should consider the factors within different environment. The CPU is scarce resource in distributed environment so that we should carefully tune the `num_parallel_calls`  instead of paying attention only to the `map_fn` processing time. @fanshiqing is preparing code in recent days. Any comments is appreciated.

**Comment by jsimsa** [comment]
@fanshiqing 

Re (1): `tf.data.Dataset.prefetch()` buffer size auto-tuning is indeed one-way but this autotuning is currently not related to the performance model related autotuning of parallelism. `tf.data.Dataset.interleave()` and `tf.data.Dataset.map()` parallelism auto-tuning is not one-way. As you pointed out, this mechanism is meant to help users create pipelines that are expected to work across different environments (and data patterns). Adding support for default value provided by the user is certainly something that we could consider but at the moment we do not see need for it.

Re (2): We plan to soon introduce a mechanism to allow users to configure the CPU budget to be used by the performance modeling for the parallelism auto-tuning.

In other words, we are aware of the limitations you have pointed out and are working on addressing some of them. Since your request does not describe a concrete issue and steps how to reproduce it, I am going to close this issue.

**Comment by yangjunpro** [comment]
@jsimsa 

Hi Jiri,

Could please kindly wait for a while to close this issue? Or at least wait for our confirmation?

As mentioned in this post, we are already working on some PR preparation based on our observation. 
If it is possible, any deeper interaction would be highly appreciated. Otherwise, we may do some duplicated work. That's why we raise what we observe as early as possible. 

Based on your closing this issue so quickly without any specific input, frankly speaking, we are not sure whether we should go on with our work about the auto-tuning enhancement or not. 

I think if Google wants to get more input from the community, maybe the interaction could be deeper.

Thanks

**Comment by jsimsa** [comment]
@yangjunpro issues should be created for bugs or feature requests.

Bugs should be accompanied by instructions on how to reproduce and feature requests should identify the desired functionality to be developed.

As far as I can tell, this issue is neither a bug nor a feature request. You are welcome to reopen this issue if you encounter a (performance) bug and can share steps on how to reproduce it. If instead you prefer to address what you believe is the limitation yourself, you can create a PR with your contribution when it is ready.

**Comment by jsimsa** [comment]
Please see https://www.tensorflow.org/community/ for details.

**Comment by fanshiqing** [comment]
@jsimsa Thanks for your comments. According to your reply, will the consumer speed (namely subsequent training threads) be considered in your performance model in development, or the performance model is just kept unchanged like this now. As feasible negative optimization will actually slow down the consumer's speed in my opinion.

**Comment by yangjunpro** [comment]
@jsimsa 

We have already delivered several PRs to the community so we are familiar with the basic flow.

From my perspective, I think Shiqing already mentioned that he thought this is a feature and he is working on it. I think the title of this issue could be adjusted a little bit to avoid misleading. 

As to the "reproduce it" mentioned by you, the limitation is observed in our in-house cluster environment, I think it is not easy to reproduce it in Google' cluster environment. 
The logic regarding this issue is straightforward when we allocate CPU resources for the IO threads, it may consume too much to block the execution of the execution/communication threads. 
We already started to work on this improvement.  If Google is also already working on it, it's fine for us to move to another target to save RD resource. But I think before making any decision, deeper and more effective communication to ensure the community and Google are at the same page. 

Thanks.

**Comment by jsimsa** [comment]
ok

**Comment by fanshiqing** [comment]
@jsimsa Thanks for your detailed reply, initial PR will be proposed within the next 3-4 days.

**Comment by fanshiqing** [comment]
@jsimsa I'm very sorry to inform you that the initial PR will be delayed compared to the plan due to other high priority tasks and Chinese National Day holiday. And I will come back to you as soon as this PR is ready.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.
